### ISSUE TITLE
AVR: Wrong program address in disassembly

### ISSUE THREAD

**Issue by Milo-D** [radareorg/radare2#18174]
## Environment

Di 5. Jan 06:23:22 CET 2021
radare2 4.3.1 0 @ linux-x86-64 git.4.3.1
commit: unknown build: 
Linux x86_64

## Description

When disassembling an AVR binary (in this case for the ATmega328), the addresses in disassembly are incremented by 2 if it's a 16-bit instruction, and by 4 if it's a 32-bit instruction, although the AVR Flash is organized as (n * 16-bit) memory.

**Expected:** An address increment of 1 if it's a 16-bit opcode and an increment of 2 if it's a 32-bit opcode.

## Test

```console
r2 main.bin
```
, where main.bin is an AVR binary.

![r2bug](https://user-images.githubusercontent.com/46600932/103610992-9d822800-4f21-11eb-824b-d82a163704f8.png)

Atmega328 contains a 32kBytes Flash which has a range from 0x0000 to 0x3fff. This is only possible if an address represents 2 bytes of memory (0x4000 * 2 = 0x8000, so 32kBytes).

![flash](https://user-images.githubusercontent.com/46600932/103611030-b094f800-4f21-11eb-85c2-917650749aed.png)

**Comment by karliss** [comment]
Even though some parts of documentation show addresses in words. SPM and LPM instructions which access flash memory use byte based addressing. So it's possible to have constant tables which are not word aligned.

Other instructions for saving bits in instruction encoding probably encode addresses as words. But architectures with byte based addressing also do this. The disassembler doesn't need to print numbers exactly as encoded. As different instructions might have different tradeoffs in the encoding.

I assume that something like intel hex files containing AVR programs also uses byte based addressing.

Slightly off-topic. I wonder how the tooling in general (not r2) and  constant tables are handled for architectures like PIC with word size that isn't multiple of 8.

So as long as all the jumps are resolved correctly and displayed consistently I wouldn't call this wrong as much as it is different view on the same data.

That said for ease of comparison with datasheet it might be convenient to have an option for word based address display mode, just like there is an option for relative addresses.

**Comment by Milo-D** [comment]
> So as long as all the jumps are resolved correctly and displayed consistently I wouldn't call this wrong as much as it is different view on the same data.

Yes, you are right, I assumed that the addressing is implemented as word based addressing ^^ I will edit the issue's title, because, if it's intended as byte address, then there is no real issue (atleast no bug) here. 

> That said for ease of comparison with datasheet it might be convenient to have an option for word based address display mode, just like there is an option for relative addresses.

That would be a good idea. So this issue probably belongs to "Feature Request", but my life is not depending on this feature, so others have to decide ^^ 

Thank you for your respond.
